mkcloud: watch out for incomplete LVM device filter

aspiers opened the pull request:
mkcloud creates LVs for use as DRBD data disks, which are then used by the controller VMs as PVs to build drbd VGs on top of.  However we need to prevent the host from detecting the LVs inside the resulting drbd VGs, because they only concern the controller VM guests.  This requires tweaking `lvm.conf` which is not easy to automate, but we can at least check whether it was done right, and bail if not.

@jdsn This was based on your suggestion for editing `lvm.conf`, but I still don't understand it 100% because it will still detect the `/dev/dm-$i` devices corresponding to these LVs, so what does it really achieve by excluding the `/dev/$vg` and `/dev/mapper/$vg-*` devices?

bmwiedemann commented:
btw: this is mostly a problem on systemd-based hosts, where udev/systemd is autoactivating too much.

on our SLES12 mkcloud host we added to the filter list:
"r|/dev/mkcvg0/.*|", "r|/dev/mapper/mkcvg0-.*|", "r|/dev/dm-*|"

bmwiedemann commented on scripts/mkcloud:
```diff
@@ -386,6 +386,11 @@ function onhost_create_cloud_lvm()
+
+    echo "Checking for LVs treated by LVM as valid PV devices ..."
+    if lvmdiskscan | egrep "/dev/($cloud/|mapper/$cloud-)"; then
+        complain 94 "Error: your lvm.conf is not filtering out mkcloud LVs; please fix."
```
could link to details how to fix.

jdsn commented:
I'd rather grep for the proper filter in lvm.conf instead of the output of lvmdiskscan.

jdsn commented:
Also needed on the filter list:
```
"r|/dev/.*/by-id/.*|", "r|/dev/block.*|"
```
They should be there by default. But we need to check them in this case as well. Somebody might have custom changes to his lvm.conf

bmwiedemann commented:
actually, people could have a working lvm.conf without these parts by just having a limited allow rule without any reject rules

aspiers commented:
@bmwiedemann is right - @jdsn I definitely disagree about grepping for filters because we care about the correct result, not the way we got there.  But thanks for the tips on other things to filter, I will improve this PR shortly!

jdsn commented:
@aspiers Ok, if the error message contians the steps needed for a cleanup then I would also agree.

aspiers commented:
@jdsn `r|/dev/block.*|` is not there by default on my 13.1 system.  Is that new in 13.2?

@bmwiedemann If `r|/dev/md-[0-9]+|` risks catching too much then what do we do?

aspiers commented:
I've rebased, added checks for LVM physical volumes represented by `/dev/md-[0-9]+`, and made the error message more helpful.  But I'm still not convinced it's correct yet.

aspiers commented:
Due to the uncertainty regarding `/dev/md-[0-9]+`, I've dropped those bits.  So let's merge this (which is already better than nothing) whilst we continue to think about how to handle those right.

jdsn commented:
Could you just add a short comment about the specialty of the /dev/dm-[0-9]+ devices to the error output? Then I will +1 it.

aspiers commented:
Sounds reasonable but I'm not sure what to write exactly.  Suggestions?

aspiers commented:
@jdsn ping...

jdsn commented:
Something like
```
The filter should also include something like "/dev/dm-[0-9]+", but the the values depend on you local system setup. Please add/modify it manually.
```

aspiers commented:
@jdsn But how would you determine the correct filter for any given system?  Surely the `dm` numbers are not stable, so there is no correct answer to this?

jdsn commented:
@aspiers the comment I am asking for does not need to be the perfect answer. It should just point out that there most likely might be more to take care of.
Just knowingly concealing it is not nice.

aspiers commented:
@jdsn Done.

jdsn commented:
thanks, :+1:

bmwiedemann commented:
:+1:
